lwgenerate/gcp/gcp.go

```diff
@@ -0,0 +1,505 @@
+
+// ExistingLogSinkName Set the Topic ARN of an existing AuditLog setup
+func ExistingLogSinkName(name string) GcpTerraformModifier {
+	return func(c *GenerateGcpTfConfigurationArgs) {
+		c.ExistingLogSinkName = name
+	}
+}
+
+// EnableForceDestroyBucket Enable force destroy of the bucket if it has stuff in it
+func EnableForceDestroyBucket() GcpTerraformModifier {
+	return func(c *GenerateGcpTfConfigurationArgs) {
+		c.EnableForceDestroyBucket = true
+	}
+}
+
+// EnableUBLA Enable force destroy of the bucket if it has stuff in it
+func EnableUBLA() GcpTerraformModifier {
+	return func(c *GenerateGcpTfConfigurationArgs) {
+		c.EnableUBLA = true
+	}
+}
+
+// LogBucketLifecycleRuleAge Set the number of days to keep audit logs in Lacework GCS bucket before deleting
+// Defaults to -1. Leave default to keep indefinitely.
+func LogBucketLifecycleRuleAge(age int) GcpTerraformModifier {
+	return func(c *GenerateGcpTfConfigurationArgs) {
+		c.LogBucketLifecycleRuleAge = age
+	}
+}
+
+// LogBucketRetentionDays Set the number of days to keep logs before deleting. Default is 30
+func LogBucketRetentionDays(days int) GcpTerraformModifier {
+	return func(c *GenerateGcpTfConfigurationArgs) {
+		c.LogBucketRetentionDays = days
+	}
+}
+
+// UseAuditLogIntegrationName Set the Config Integration name to be displayed on the Lacework UI
+func UseAuditLogIntegrationName(name string) GcpTerraformModifier {
+	return func(c *GenerateGcpTfConfigurationArgs) {
+		c.AuditLogIntegrationName = name
+	}
+}
+
+// Generate new Terraform code based on the supplied args.
+func (args *GenerateGcpTfConfigurationArgs) Generate() (string, error) {
+	// Validate inputs
+	if err := args.validate(); err != nil {
+		return "", errors.Wrap(err, "invalid inputs")
+	}
+
+	// Create blocks
+	requiredProviders, err := createRequiredProviders()
+	if err != nil {
+		return "", errors.Wrap(err, "failed to generate required providers")
+	}
+
+	gcpProvider, err := createGcpProvider(args)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to generate gcp provider")
+	}
+
+	laceworkProvider, err := createLaceworkProvider(args)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to generate lacework provider")
+	}
+
+	configModule, err := createConfig(args)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to generate aws config module")
+	}
+
+	auditLogModule, err := createAuditLog(args)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to generate gcp audit log module")
+	}
+
+	// Render
+	hclBlocks := lwgenerate.CreateHclStringOutput(
+		lwgenerate.CombineHclBlocks(
+			requiredProviders,
+			gcpProvider,
+			laceworkProvider,
+			configModule,
+			auditLogModule),
+	)
+	return hclBlocks, nil
+}
+
+func createRequiredProviders() (*hclwrite.Block, error) {
+	return lwgenerate.CreateRequiredProviders(
+		lwgenerate.NewRequiredProvider("lacework",
+			lwgenerate.HclRequiredProviderWithSource("lacework/lacework"),
+			lwgenerate.HclRequiredProviderWithVersion("~> 0.12.2")))
```

Comment:
Somehow I feel we need a single place that has all of these versions defined as constants so that we don't have to go find them deep in the codebase when we have to do a major version bump.

Reply:
Wondering if we should create a constants file at the top level of the `lwgenerate` folder and set the versions and provider sources there?

Reply:
Created an `lwgenerate` package `constants.go` and updated :)